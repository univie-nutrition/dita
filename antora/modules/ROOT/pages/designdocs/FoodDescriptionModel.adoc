= Food Description Model
:includedir: includes

In the following chapters we lay out how food is _described_.

include::{includedir}/SemanticIdentifier.adoc[leveloffset=1]

== Classification and Description

The _Food Description Model_ provides those building blocks, necessary to describe food intake. 
Recipes are composed of ingredients, with each ingredient being quantified, described and having additional coefficients
to account for non-edibel part losses and nutrient losses during the cooking process.     

Food is _classified_ and _described_ by

* Food Groups, Subgroups and Sub-Subgroups
* Food Facets and Descriptors footnote:id[as per interview (`Food` and `Recipe` have no descriptors per-se, as those are subject to intake surveying. Only `Ingredients` do have those.)]

Recipes are _classified_ and _described_ by

* Recipe Groups and Subgroups
* Recipe Facets and Descriptors footnote:id[]
* Ingredients

[plantuml,fig-food-description-model,svg]
.Food Description Model
----
@startuml

class Food {
    sid: SemanticIdentifier 
    name: String
    attributes: Map<String, String>
}

class Recipe {
    sid: SemanticIdentifier
    name: String
    attributes: Map<String, String>
}

class RecipeIngredient {
    amountGrams: BigDecimal 
    rawToCookedRatio: BigDecimal
}

class ClassificationFacet {
    sid: SemanticIdentifier 
    name: String 
}

RecipeIngredient --> Recipe : recipeSid (parent)
RecipeIngredient --> Food : foodSid
Food --> ClassificationFacet : groupSid
Recipe --> ClassificationFacet : groupSid
RecipeIngredient --> "*" ClassificationFacet : foodFacetSids

note right of ClassificationFacet
  unifies groups 
  and descriptors
end note

hide class method

@enduml
----
