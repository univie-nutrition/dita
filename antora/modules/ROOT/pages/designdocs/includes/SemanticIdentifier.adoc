= Semantic Identifier

A _Semantic Identifier_ (SID) references data objects across system boundaries.

[plantuml,fig-sid,svg]
.Semantic Identifier (SID)
----
@startuml

class SystemId {
    system: String
    version: String
}

class ObjectId {
    context: String
    objectSimpleId: String
}

class SemanticIdentifier {
}

class SemanticIdentifierSet {
}

note right of SemanticIdentifier::systemId
  identifies the system (and optionally system version) of the referenced data object
end note
note right of SemanticIdentifier::objectId
  uniquely identifies the data object within the system
end note

SemanticIdentifier --> SystemId
SemanticIdentifier --> ObjectId

SemanticIdentifierSet -right-> "*" SemanticIdentifier : has many

hide class method

@enduml
----

Example use, referencing a food, product or facet.

[source,java]
.Java
----
// within BLS v3.02 represents 'Banana raw'
var blsBanana = SemanticIdentifier.parse("de.bls/3.02", "food/F503100");
var totalSuguar = SemanticIdentifier.parse("de.bls/3.02", "comp/KMD");

// within LanguaL 2017 Facet Category - A (PRODUCT TYPE) represents 'Bananas' from GS1 ontology
var gs1BananaId = new SemanticIdentifier("langual/2017:A1415"); 
----