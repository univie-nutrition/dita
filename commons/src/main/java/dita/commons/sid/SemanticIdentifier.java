/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.commons.sid;

import java.util.function.UnaryOperator;

import org.springframework.lang.Nullable;

import org.apache.causeway.commons.internal.base._Strings;

import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.experimental.Accessors;

/**
 * A Semantic Identifier references data objects across system boundaries.
 * <p>
 * <ul>
 * <li>system: at.gd</li>
 * <li>version: 2.0</li>
 * <li>context: food | recipe | food.desc | recipe.desc | food.group etc.</li>
 * <li>object: 00123</li>
 * </ul>
 * <p>
 * formats to -> SID[at.gd/2.0, food/00123]
 */
@Builder
public record SemanticIdentifier (
        /**
         * Identifies the system of the referenced data object.
         * (optionally includes version information)
         */
        @NonNull SystemId systemId,
        /**
         * Uniquely identifies the data object within the system.
         * (optionally includes context information)
         */
        @NonNull ObjectId objectId) implements Comparable<SemanticIdentifier> {

    // -- SYSTEM ID

    public record SystemId(
            /**
             * Identifies the system of the referenced data object.
             */
            String system,
            /**
             * Identifies the system's version of the referenced data object.
             */
            String version) implements Comparable<SystemId> {

        public static SystemId empty() {
            return new SystemId(null, null);
        }

        public static SystemId parse(final @Nullable String systemIdStringified) {
            return _Utils.parseSystemId(systemIdStringified);
        }

        public SystemId(final String system, final String version) {
            this.system = _Utils.validate(system);
            this.version = _Utils.validate(version);
        }

        /**
         * SystemId constructor with empty version part.
         */
        public SystemId(final String system) {
            this(system, null);
        }

        // -- WITHER

        public SystemId withSystem(final String system) {
            return new SystemId(system, version());
        }
        public SystemId withVersion(final String version) {
            return new SystemId(system(), version);
        }

        // -- MAPPER

        public SystemId mapSystem(final UnaryOperator<String> systemMapper) {
            return withSystem(systemMapper.apply(system()));
        }
        public SystemId mapVersion(final UnaryOperator<String> versionMapper) {
            return withVersion(versionMapper.apply(version()));
        }

        // -- CONTRACT

        @Override
        public int compareTo(final @Nullable SystemId o) {
            return compare(this, o);
        }

        /**
         * Identifies the system of the referenced data object.
         * (optionally suffixed with a version)
         */
        @Override
        public final String toString() {
            return _Utils.format(this);
        }

        public static int compare(
                final @Nullable SystemId a,
                final @Nullable SystemId b) {
            if(a==null) return b==null
                        ? 0
                        : -1;
            if(b==null) return 1;
            final int c = _Strings.compareNullsFirst(a.system(), b.system());
            return c!=0
                ? c
                // TODO should use DEWEY compare here
                : _Strings.compareNullsFirst(a.version(), b.version());
        }

    }

    // -- OBJECT ID

    public record ObjectId(
            /**
             * Uniquely identifies the data object's context within the system.
             */
            String context,
            /**
             * Uniquely identifies the data object within the system's context.
             */
            String object) implements Comparable<ObjectId> {

        /**
         * Some predefined contexts.
         */
        @RequiredArgsConstructor
        public enum Context {
            LITERAL("lit"),
            LANGUAGE("lang"),
            COMPONENT("comp"),
            FOOD("food"),
            RECIPE("recp"),
            BRAND("brand"),
            FOOD_DESCRIPTOR("fd"),
            RECIPE_DESCRIPTOR("rd"),
            FOOD_GROUP("fg"),
            RECIPE_GROUP("rg"),
            ;
            @Getter @Accessors(fluent=true)
            final String id;
        }

        public static ObjectId empty() {
            return new ObjectId((String)null, null);
        }

        public static ObjectId parse(final @Nullable String objectIdStringified) {
            return _Utils.parseObjectId(objectIdStringified);
        }

        public ObjectId(final String context, final String object) {
            this.context = _Utils.validate(context);
            this.object = _Utils.validateObject(object);
        }

        public ObjectId(final @NonNull Context context, final String object) {
            this(context.id, object);
        }

        /**
         * ObjectId constructor with empty context part.
         */
        public ObjectId(final String object) {
            this((String)null, object);
        }

        // -- WITHER

        public ObjectId withContext(final String context) {
            return new ObjectId(context, object());
        }
        public ObjectId withObject(final String object) {
            return new ObjectId(context(), object);
        }

        // -- MAPPER

        public ObjectId mapContext(final UnaryOperator<String> contextMapper) {
            return withContext(contextMapper.apply(context()));
        }
        public ObjectId mapObject(final UnaryOperator<String> objectMapper) {
            return withObject(objectMapper.apply(object()));
        }

        // -- CONTRACT

        @Override
        public int compareTo(final @Nullable ObjectId o) {
            return compare(this, o);
        }

        /**
         * Uniquely identifies the data object within the system.
         * (optionally prefixed with a context)
         */
        @Override
        public final String toString() {
            return _Utils.format(this);
        }

        public static int compare(
                final @Nullable ObjectId a,
                final @Nullable ObjectId b) {
            if(a==null) return b==null
                        ? 0
                        : -1;
            if(b==null) return 1;
            final int c = _Strings.compareNullsFirst(a.context(), b.context());
            return c!=0
                ? c
                // TODO should use DEWEY compare here
                : _Strings.compareNullsFirst(a.object(), b.object());
        }

    }

    // -- SID IMPL

    public static SemanticIdentifier empty() {
        return new SemanticIdentifier(SystemId.empty(), ObjectId.empty());
    }

    public SemanticIdentifier(final String systemId, final String objectId) {
        this(SystemId.parse(systemId), ObjectId.parse(objectId));
    }

    // -- PARSER

    /**
     * Inverse of {@link SemanticIdentifier#toString()}.
     * @see SemanticIdentifier#toString()
     */
    public static SemanticIdentifier parse(final @Nullable String stringified) {
        return _Utils.parseSid(stringified);
    }

    // -- WITHER

    public SemanticIdentifier withSystemId(final SystemId systemId) {
        return new SemanticIdentifier(systemId, objectId());
    }
    public SemanticIdentifier withObjectId(final ObjectId objectId) {
        return new SemanticIdentifier(systemId(), objectId);
    }

    // -- MAPPER

    public SemanticIdentifier mapSystemId(final UnaryOperator<SystemId> systemIdMapper) {
        return withSystemId(systemIdMapper.apply(systemId()));
    }
    public SemanticIdentifier mapObjectId(final UnaryOperator<ObjectId> objectIdMapper) {
        return withObjectId(objectIdMapper.apply(objectId()));
    }


    @Override
    public int compareTo(final @Nullable SemanticIdentifier o) {
        return compare(this, o);
    }

    /**
     * e.g. {@code at.gd/2.0:food/00123}
     */
    public String toStringNoBox() {
        return _Utils.formatUnboxed(this);
    }

    /**
     * e.g. {@code SID[at.gd/2.0:food/00123]}
     */
    @Override
    public final String toString() {
        return _Utils.formatBoxed(this);
    }

    // -- UTILITY

    public static int compare(
            final @Nullable SemanticIdentifier a,
            final @Nullable SemanticIdentifier b) {
        if(a==null) return b==null
                    ? 0
                    : -1;
        if(b==null) return 1;
        final int c = a.systemId().compareTo(b.systemId());
        return c!=0
            ? c
            : a.objectId().compareTo(b.objectId());
    }

}
