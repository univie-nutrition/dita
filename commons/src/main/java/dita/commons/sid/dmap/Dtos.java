/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.commons.sid.dmap;

import java.util.Collection;
import java.util.concurrent.ConcurrentHashMap;

import org.jspecify.annotations.Nullable;

import org.apache.causeway.commons.functional.Try;
import org.apache.causeway.commons.internal.exceptions._Exceptions;
import org.apache.causeway.commons.io.DataSource;
import org.apache.causeway.commons.io.YamlUtils;
import org.apache.causeway.commons.io.YamlUtils.YamlLoadCustomizer;

import lombok.experimental.UtilityClass;

import dita.commons.util.FormatUtils;

@UtilityClass
class Dtos {

    // -- DIRECT MAP

    public record DirectMapDto(
            Collection<DirectMapEntry> directMapEntries) {
        public String toYaml() {
            return YamlUtils.toStringUtf8(this, FormatUtils.yamlOptions());
        }
        public static Try<DirectMapDto> tryFromYaml(@Nullable final DataSource ds) {
            if(ds==null) return Try.failure(_Exceptions.noSuchElement("missing datasource"));
            return YamlUtils.tryReadCustomized(
                    DirectMapDto.class, ds, yamlMillionCodePointsLimit(200), FormatUtils.yamlOptions());
        }
    }

    DirectMapDto toDto(final DirectMap map) {
        return new DirectMapDto(
                map.streamEntries()
                    .toList());
    }

    DirectMap fromDto(@Nullable final DirectMapDto dto) {
        if(dto==null) return null;
        var map = new DirectMap(new ConcurrentHashMap<>());
        dto.directMapEntries.stream()
            .forEach(map::put);
        return map;
    }

    // -- HELPER

    private YamlLoadCustomizer yamlMillionCodePointsLimit(final int millions) {
        return loader->{
            loader.setCodePointLimit(millions * 1_000_000);
            return loader;
        };
    }

}
