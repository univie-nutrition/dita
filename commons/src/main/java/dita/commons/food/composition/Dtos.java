/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.commons.food.composition;

import java.math.BigDecimal;
import java.util.Collection;
import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.lang.Nullable;

import org.apache.causeway.commons.functional.Try;
import org.apache.causeway.commons.internal.exceptions._Exceptions;
import org.apache.causeway.commons.io.DataSource;
import org.apache.causeway.commons.io.JsonUtils.JacksonCustomizer;
import org.apache.causeway.commons.io.YamlUtils;
import org.apache.causeway.commons.io.YamlUtils.YamlLoadCustomizer;

import lombok.NonNull;
import lombok.experimental.UtilityClass;

import dita.commons.food.composition.FoodComponentDatapoint.DatapointSemantic;
import dita.commons.food.composition.FoodComposition.CompositionQuantification;
import dita.commons.jaxb.JaxbAdapters;
import dita.commons.sid.SemanticIdentifier;

@UtilityClass
class Dtos {

    // -- DATAPOINT

    public record FoodComponentDatapointDto(
            @NonNull SemanticIdentifier componentId,
            @NonNull DatapointSemantic datapointSemantic,
            @NonNull BigDecimal datapointValue) {
    }

    FoodComponentDatapointDto toDto(final FoodComponentDatapoint datapoint) {
        return new FoodComponentDatapointDto(
                datapoint.component().componentId(),
                datapoint.datapointSemantic(),
                datapoint.datapointValue());
    }

    FoodComponentDatapoint fromDto(
            final FoodComponentDatapointDto dto,
            final CompositionQuantification compositionQuantification,
            final FoodComponentCatalog componentCatalog) {
        return new FoodComponentDatapoint(
                componentCatalog.lookupEntryElseFail(dto.componentId()),
                compositionQuantification,
                dto.datapointSemantic(),
                dto.datapointValue());
    }

    // -- FOOD COMPOSITION

    public record FoodCompositionDto(
            @NonNull SemanticIdentifier foodId,
            @NonNull CompositionQuantification compositionQuantification,
            @NonNull Collection<FoodComponentDatapointDto> datapoints) {
    }

    FoodCompositionDto toDto(final FoodComposition composition) {
        return new FoodCompositionDto(
                composition.foodId(),
                composition.compositionQuantification(),
                composition.streamDatapoints().map(Dtos::toDto).toList());
    }

    FoodComposition fromDto(
            final FoodCompositionDto dto,
            final FoodComponentCatalog componentCatalog) {
        var datapointMap = new HashMap<SemanticIdentifier, FoodComponentDatapoint>();
        dto.datapoints().stream()
            .map(d->Dtos.fromDto(d, dto.compositionQuantification(), componentCatalog))
            .forEach(dp->datapointMap.put(dp.componentId(), dp));
        return new FoodComposition(dto.foodId(), dto.compositionQuantification(), datapointMap);
    }

    // -- FOOD COMPOSITION REPOSITORY

    public record FoodCompositionRepositoryDto(
            Collection<FoodComponent> components,
            Collection<FoodCompositionDto> compositions) {

        public String toYaml() {
            return YamlUtils.toStringUtf8(this, yamlOptions());
        }
        public static Try<FoodCompositionRepositoryDto> tryFromYaml(@Nullable final DataSource ds) {
            if(ds==null) return Try.failure(_Exceptions.noSuchElement("missing datasource"));
            return YamlUtils.tryReadCustomized(
                    FoodCompositionRepositoryDto.class, ds, yamlMillionCodePointsLimit(200), yamlOptions());
        }
    }

    FoodCompositionRepositoryDto toDto(final FoodCompositionRepository repo) {
        return new FoodCompositionRepositoryDto(
                repo.componentCatalog().streamComponents().toList(),
                repo.streamCompositions().map(Dtos::toDto).toList());
    }

    FoodCompositionRepository fromDto(@Nullable final FoodCompositionRepositoryDto dto) {
        if(dto==null) return null;
        var componentCatalog = new FoodComponentCatalog();
        var internalMap = new ConcurrentHashMap<SemanticIdentifier, FoodComposition>();
        dto.components().forEach(componentCatalog::put);
        dto.compositions().forEach(comp->internalMap.put(comp.foodId(), Dtos.fromDto(comp, componentCatalog)));
        var repo = new FoodCompositionRepository(componentCatalog, internalMap);
        return repo;
    }

    // -- HELPER

    private JacksonCustomizer yamlOptions() {
        return JacksonCustomizer.wrapXmlAdapter(new JaxbAdapters.SemanticIdentifierAdapter());
    }

    private YamlLoadCustomizer yamlMillionCodePointsLimit(final int millions) {
        return loader->{
            loader.setCodePointLimit(millions * 1_000_000);
            return loader;
        };
    }

}
