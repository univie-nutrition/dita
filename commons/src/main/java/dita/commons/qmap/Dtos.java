/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.commons.qmap;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.lang.Nullable;

import org.apache.causeway.commons.functional.Try;
import org.apache.causeway.commons.internal.exceptions._Exceptions;
import org.apache.causeway.commons.io.DataSource;
import org.apache.causeway.commons.io.YamlUtils;
import org.apache.causeway.commons.io.YamlUtils.YamlLoadCustomizer;

import lombok.NonNull;
import lombok.experimental.UtilityClass;

import dita.commons.format.FormatUtils;
import dita.commons.qmap.QualifiedMap.Policy;
import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifierSet;

@UtilityClass
class Dtos {

    // -- QUALIFIED MAP ENTRY

    public record QualifiedMapEntryDto(
            @NonNull SemanticIdentifier source,
            @Nullable List<SemanticIdentifier> qualifier,
            @Nullable SemanticIdentifier target) {
    }

    QualifiedMapEntryDto toDto(final QualifiedMapEntry entry) {
        return new QualifiedMapEntryDto(
                entry.source(),
                entry.qualifier()!=null
                    ? entry.qualifier().elements().toList()
                    : null,
                entry.target());
    }

    QualifiedMapEntry fromDto(final QualifiedMapEntryDto dto) {
        return new QualifiedMapEntry(
                dto.source,
                SemanticIdentifierSet.ofCollection(dto.qualifier),
                dto.target);
    }

    // -- QUALIFIED MAP

    public record QualifiedMapDto(
            Collection<QualifiedMapEntryDto> qualifiedMapEntries) {
        public String toYaml() {
            return YamlUtils.toStringUtf8(this, FormatUtils.yamlOptions());
        }
        public static Try<QualifiedMapDto> tryFromYaml(@Nullable final DataSource ds) {
            if(ds==null) return Try.failure(_Exceptions.noSuchElement("missing datasource"));
            return YamlUtils.tryReadCustomized(
                    QualifiedMapDto.class, ds, yamlMillionCodePointsLimit(200), FormatUtils.yamlOptions());
        }
    }

    QualifiedMapDto toDto(final QualifiedMap map) {
        return new QualifiedMapDto(
                map.streamEntries()
                    .map(Dtos::toDto)
                    .toList());
    }

    QualifiedMap fromDto(@Nullable final QualifiedMapDto dto, Policy policy) {
        if(dto==null) return null;
        var map = new QualifiedMap(new ConcurrentHashMap<>(), policy);
        dto.qualifiedMapEntries.stream()
            .map(Dtos::fromDto)
            .forEach(map::put);
        return map;
    }
    
    // -- HELPER

    private YamlLoadCustomizer yamlMillionCodePointsLimit(final int millions) {
        return loader->{
            loader.setCodePointLimit(millions * 1_000_000);
            return loader;
        };
    }

}
