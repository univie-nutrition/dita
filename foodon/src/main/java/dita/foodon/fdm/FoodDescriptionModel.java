/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.foodon.fdm;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;

import org.jspecify.annotations.Nullable;

import org.apache.causeway.commons.internal.base._NullSafe;
import org.apache.causeway.commons.internal.base._Strings;
import org.apache.causeway.commons.internal.exceptions._Exceptions;
import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifierSet;

/**
 * Provides a set of food ontologies,
 * that allow to describe consumed food,
 * including categorization (grouping) and descriptive facets.
 */
public record FoodDescriptionModel(
        Map<SemanticIdentifier, Food> foodBySid,
        Map<SemanticIdentifier, Recipe> recipeBySid,
        Map<SemanticIdentifier, List<RecipeIngredient>> ingredientsByRecipeSid,
        Map<SemanticIdentifier, ClassificationFacet> classificationFacetBySid
        ) {

    public sealed interface HasAttributes
    permits FoodDescriptionModel.Food, FoodDescriptionModel.Recipe {
        Map<String, String> attributes();
        /// @returns optionally the attribute value corresponding to given attributeName
        default Optional<String> attribute(final String attributeName) {
            return _Strings.nonEmpty(attributes().get(attributeName));
        }
    }

    public record Food(
            SemanticIdentifier sid,
            String name,
            SemanticIdentifier groupSid,
            Map<String, String> attributes) implements HasAttributes {
    }

    public record Recipe(
            SemanticIdentifier sid,
            String name,
            SemanticIdentifier groupSid,
            Map<String, String> attributes) implements HasAttributes {
    }

    public record RecipeIngredient(
            SemanticIdentifier recipeSid,
            SemanticIdentifier foodSid,
            SemanticIdentifierSet foodFacetSids,
            /**
             * Amount consumed in gram.
             */
            BigDecimal amountGrams,
            BigDecimal rawToCookedRatio) {
    }

    public record ClassificationFacet(
            SemanticIdentifier sid,
            String name) {
    }

    // -- FACTORIES

    public static FoodDescriptionModel empty() {
        return new FoodDescriptionModel(
                new HashMap<SemanticIdentifier, Food>(),
                new HashMap<SemanticIdentifier, Recipe>(),
                new HashMap<SemanticIdentifier, List<RecipeIngredient>>(),
                new HashMap<SemanticIdentifier, ClassificationFacet>());
    }

    // -- UTIL

    public Optional<Food> lookupFoodBySid(@Nullable final SemanticIdentifier sid) {
        return sid!=null
                ? Optional.ofNullable(foodBySid().get(sid))
                : Optional.empty();
    }

    public Optional<Recipe> lookupRecipeBySid(@Nullable final SemanticIdentifier sid) {
        return sid!=null
                ? Optional.ofNullable(recipeBySid().get(sid))
                : Optional.empty();
    }

    public Food lookupFoodBySidElseFail(@Nullable final SemanticIdentifier sid) {
        return lookupFoodBySid(sid)
                .orElseThrow(()->_Exceptions.illegalArgument("failed to resolve food for sid %s", sid));
    }

    public Recipe lookupRecipeBySidElseFail(@Nullable final SemanticIdentifier sid) {
        return lookupRecipeBySid(sid)
                .orElseThrow(()->_Exceptions.illegalArgument("failed to resolve recipe for sid %s", sid));
    }

    /**
     * Streams all ingredients of given recipe.
     */
    public Stream<RecipeIngredient> streamIngredients(@Nullable final Recipe recipe) {
        return recipe!=null
                ? _NullSafe.stream(ingredientsByRecipeSid.get(recipe.sid()))
                : Stream.empty();
    }

    /**
     * Returns sum of amount gram consumed over all ingredients of given recipe.
     */
    public BigDecimal sumAmountGramsForRecipe(@Nullable final Recipe recipe) {
        return streamIngredients(recipe)
            .map(RecipeIngredient::amountGrams)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    // -- IO

    public static FoodDescriptionModel fromYaml(final String yaml) {
        return FdmUtils.fromYaml(yaml);
    }

    public String toYaml() {
        return FdmUtils.toYaml(this);
    }

}
