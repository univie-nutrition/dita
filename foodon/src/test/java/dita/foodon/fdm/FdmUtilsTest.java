/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.foodon.fdm;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Stream;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifierSet;
import dita.foodon.fdm.Dtos.FoodDescriptionModelDto;
import dita.foodon.fdm.FoodDescriptionModel.ClassificationFacet;
import dita.foodon.fdm.FoodDescriptionModel.Food;
import dita.foodon.fdm.FoodDescriptionModel.Recipe;
import dita.foodon.fdm.FoodDescriptionModel.RecipeIngredient;
import dita.testing.ApprovalTestOptions;

class FdmUtilsTest {

    @Test
    @UseReporter(DiffReporter.class)
    void toYaml() {
        var dto = FoodDescriptionModelDto.builder()
                .classificationFacets(List.of(
                        new ClassificationFacet(SemanticIdentifier.parse("test:fd/A"), "Food Descriptor A"),
                        new ClassificationFacet(SemanticIdentifier.parse("test:fd/B"), "Food Descriptor B"),
                        new ClassificationFacet(SemanticIdentifier.parse("test:fg/01"), "Fruits"),
                        new ClassificationFacet(SemanticIdentifier.parse("test:fg/02"), "Vegetables"),
                        new ClassificationFacet(SemanticIdentifier.parse("test:rg/01"), "Common Recipes")
                        ))
                .food(List.of(
                        new Food(SemanticIdentifier.parse("test:food/0001"), "Banana",
                                SemanticIdentifier.parse("test:fg/01")),
                        new Food(SemanticIdentifier.parse("test:food/0002"), "Apple",
                                SemanticIdentifier.parse("test:fg/01"))))
                .recipes(List.of(
                        new Recipe(SemanticIdentifier.parse("test:recipe/0002"), "Spaghetti",
                                SemanticIdentifier.parse("test:rg/01"))
                        ))
                .ingredients(List.of(
                        new RecipeIngredient(SemanticIdentifier.parse("test:recipe/0002"),
                                SemanticIdentifier.parse("test:food/0002"),
                                SemanticIdentifierSet.ofStream(
                                        Stream.of(
                                                SemanticIdentifier.parse("test:fd/1234"),
                                                SemanticIdentifier.parse("test:fd/5678"))),
                                new BigDecimal("0.123"),
                                new BigDecimal("0.324"))
                        ))
                .build();
        var fdm = FdmUtils.fromDto(dto);
        var yaml = FdmUtils.toYaml(fdm);

        //debug
        //System.err.printf("%s%n", yaml);

        Approvals.verify(yaml, ApprovalTestOptions.yamlOptions());
        assertEquals(fdm, FdmUtils.fromYaml(yaml), ()->"roundtrip on yaml failed");
    }

}
