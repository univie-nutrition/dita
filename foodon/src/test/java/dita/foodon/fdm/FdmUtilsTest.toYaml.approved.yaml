food:
- sid: "test:food/0002"
  name: "Apple"
  groupSid: "test:fg/01"
- sid: "test:food/0001"
  name: "Banana"
  groupSid: "test:fg/01"
recipes:
- sid: "test:recipe/0002"
  name: "Spaghetti"
  groupSid: "test:rg/01"
ingredients:
- recipeSid: "test:recipe/0002"
  foodSid: "test:food/0002"
  foodFacetSids: "test:fd/1234, test:fd/5678"
  amountGrams: 0.123
  rawToCookedRatio: 0.324
classificationFacets:
- sid: "test:rg/01"
  name: "Common Recipes"
- sid: "test:fg/01"
  name: "Fruits"
- sid: "test:fg/02"
  name: "Vegetables"
- sid: "test:fd/A"
  name: "Food Descriptor A"
- sid: "test:fd/B"
  name: "Food Descriptor B"
